### PR TITLE
[Graphbolt] Add the `preprocess_ondisk_dataset` function.

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

~~I will add the corresponding **test** later.~~
- [x] Add the homogenous graph test.
- [ ] Add the heterogeneous graph test.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by keli-wen** [comment]
@dgl-bot

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -16,13 +32,183 @@
...
+                    feature["format"],
+                    in_memory=feature["in_memory"],
+                )
+                save_data(
+                    data,
+                    dataset_path / out_feature["path"],
+                    out_feature["format"],
+                )
+            elif feature["format"] == "numpy":
+                # If the original format is numpy, just copy the file.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:153]: Do we need to copy file? Can we just refer to the existing file to avoid unnecessary duplication?

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:153]: Same question to other unnecessary copies.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -16,13 +32,183 @@
...
+                intput_set_split, output_set_split
+            ):
+                # Always save the feature in numpy format.
+                output_set_per_type["format"] = "numpy"
+                output_set_per_type[
+                    "path"
+                ] = processed_dir_prefix / input_set_per_type["path"].replace(
+                    "pt", "npy"
+                )
+                if input_set_per_type["format"] != "numpy":
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:184]: Does the following conversion supports all kinds of format or just torch?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 3]
```diff
@@ -16,13 +32,183 @@
...
+            edge_data = pd.read_csv(
+                dataset_path / edge_info["path"], names=["src", "dst"]
+            )
+            src = torch.tensor(edge_data["src"])
+            dst = torch.tensor(edge_data["dst"])
+            data_dict[tuple(edge_info["type"].split(":"))] = (src, dst)
+        # Construct the heterograph.
+        g = dgl.heterograph(data_dict, num_nodes_dict)
+
+    # 3. Load the node/edge features and add them to the sampling-graph.
```
[the remark this answers is not in the document]
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:96]: Done

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 4]
```diff
@@ -16,13 +32,183 @@
...
+        The path to the input config file.
+
+    Returns
+    -------
+    output_config_path : str
+        The path to the output config file.
+    """
+    # 0. Load the input_config.
+    with open(input_config_path, "r") as f:
+        input_config = yaml.safe_load(f)
```
[the remark this answers is not in the document]
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:51]: Yes, If the input config does not contain the `graph` field, then we assume that the input config is already pre-processed. And, I'll add the comment.

[the remark this answers is not in the document]
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:51]: But, I don't know whether we need the `print`, when we find input config dose not contain the `graph` field.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 1 above)
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:153]: I've already discussed it with @Rhett-Ying. We hope hat all the processed data can be placed in a unified folder.
I'll add a TODO here.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 2 above)
Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:184]: Currently we only support `torch` and `numpy` formats.

Comment (frozenbugs) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:184]: Let's follow the pattern in # 6. It is less error prone in the future when we introduce a new format.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:51]: I have less concern about the print info.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 2 above)
Comment (frozenbugs) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:187]: Is it resolved?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 5]
```diff
@@ -16,13 +32,186 @@
...
+                else:
+                    os.makedirs(
+                        dataset_path
+                        / os.path.dirname(output_set_per_type["path"]),
+                        exist_ok=True,
+                    )
+                    shutil.copy(
+                        dataset_path / input_set_per_type["path"],
+                        dataset_path / output_set_per_type["path"],
+                    )
```
Comment (frozenbugs) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:207]:
```suggestion
                if input_set_per_type["format"] == "torch":
                    # If the original format is not numpy, convert it to numpy.
                    input_set = read_data(
                        dataset_path / input_set_per_type["path"],
                        input_set_per_type["format"],
                    )
                    save_data(
                        input_set,
                        dataset_path / output_set_per_type["path"],
                        output_set_per_type["format"],
                    )
                elif input_set_per_type["format"] == "numpy":
                    os.makedirs(
                        dataset_path
                        / os.path.dirname(output_set_per_type["path"]),
                        exist_ok=True,
                    )
                    shutil.copy(
                        dataset_path / input_set_per_type["path"],
                        dataset_path / output_set_per_type["path"],
                    )
                else:
                    raise ValueError(
                        f"Unsupported feature format {feature['format']}"
                    )
```

Comment (keli-wen) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:207]: I think we can make full use of the internal code of `read_ data`:
```
def read_data(path, fmt, in_memory=True):
    """Read data from disk."""
    if fmt == "torch":
        return _read_torch_data(path)
    elif fmt == "numpy":
        return _read_numpy_data(path, in_memory=in_memory)
    else:
        raise RuntimeError(f"Unsupported format: {fmt}")
```

I will modify the logic and only make a special judgment if `copy` is needed.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**